devel/githubContrib.go

```diff
@@ -0,0 +1,253 @@
+type UserData struct {
+	Name    string `json:"name"` // full name
+	Company string `json:"company"`
+	Email   string `json:"email"`
+}
+
+var fromEmailCount = 0 // global variable ftl (or ftw)
+
+// RemoveFromEnd returns a string without the suffix if found
+func RemoveFromEnd(s string, what string) string {
+	if !strings.HasSuffix(s, what) {
+		return s
+	}
+	return s[:len(s)-len(what)]
+}
+
+// GetCompany returns its best guess of the company for a given GitHub user login
+func GetCompany(login string, contribCount int64) string {
+	url := GitHubAPIURL("users/" + login)
+	body := GetBodyForURL(url)
+	var userData UserData
+	err := json.Unmarshal(body, &userData)
+	Check(err, "Unable to parse json")
+	var company string
+	if userData.Company != "" {
+		company = strings.Replace(userData.Company, " ", "", -1)
+	} else if userData.Email != "" {
+		fromEmailCount++
+		company = strings.Split(userData.Email, "@")[1]
+	}
+	// Many people have an @ and/or trailing spaces:
+	company = strings.ToLower(strings.TrimLeft(strings.TrimSpace(company), "@"))
+	company = RemoveFromEnd(company, ".com")
+	company = RemoveFromEnd(company, ".")
+	company = RemoveFromEnd(company, "inc")
+	company = RemoveFromEnd(company, ",")
+	company = RemoveFromEnd(company, ".")
+	// also treat gmail as unknown
+	if company != "" && company != "gmail" {
+		return strings.ToUpper(company[:1]) + company[1:]
+	}
+	log.Printf("%s (%s) has %d contributions but no company nor email", login, userData.Name, contribCount)
+	return "Unknown"
+}
+
+// --- Main --
+
+const minContributions = 3
+const debugJSON = false
+
+func main() {
+	// fmt.Println("in main")
+
+	// Get the repos for the org:
+	const org = "istio"
+	url := GitHubAPIURL("orgs/" + org + "/repos")
+	body := GetBodyForURL(url)
+	var repos []Repo
+	err := json.Unmarshal(body, &repos)
+	Check(err, "Unable to parse json")
+	log.Printf("%s has %d repos", org, len(repos))
+	if debugJSON {
+		PrettyPrintJSON(body)
+	}
+	// For each repo, get populate the user/contrib counts:
+	userMap := make(map[string]int64)
+	forksCount := 0
+	for _, r := range repos {
+		if r.IsFork {
+			log.Printf("Skipping %s which is a fork", r.Name)
+			forksCount++
+			continue
+		}
+		body := GetBodyForURL(r.ContributorsURL)
+		//PrettyPrintJSON(body)
+		var users []UserC
+		err = json.Unmarshal(body, &users)
+		Check(err, "Unable to parse json")
+		for _, u := range users {
+			userMap[u.Login] += u.Contributions
+		}
+	}
+	log.Printf("%s has %d forks", org, forksCount)
+	skippedUsers := 0
+	contributors := 0
+	// Contributor and contributions count by company
+	type CoCounts struct {
+		contributors  int
+		contributions int64
+	}
+	companiesMap := make(map[string]CoCounts)
+	for u, c := range userMap {
+		if c >= minContributions {
+			contributors++
+			fmt.Printf("user %d %s\n", c, u)
+			company := GetCompany(u, c)
+			// yuck! why is that tmp needed... because https://github.com/golang/go/issues/3117
```

Comment:
but instead of the `tmp` I would need to test for nil and allocate a cell, I think having a small by value struct is cleaner in this case no ? (just that I wish go would do the tmp behind the scenes automagically)